### ISSUE TITLE
Store intermediate files in subfolder in /tmp

### ISSUE THREAD

**Issue by dpecka** [Beep6581/RawTherapee#6189]
Hello,

RT leaves behind in /tmp directory files with 755 permissions, they could eventually fill the drive

-rw-r--r-- 1 dpecka users   35M Mar 23 02:39 DSC_8026-2.tif
-rw-r--r-- 1 dpecka users   74M Dec 23 15:10 20200125_161844.tif
-rw-r--r-- 1 dpecka users   92M Mar 26 12:19 DSC_0566.tif
-rw-r--r-- 1 dpecka users  138M Mar 25 22:58 DSC_0569.tif
-rw-r--r-- 1 dpecka users  138M Mar 26 00:10 DSC_0569-2.tif
-rw-r--r-- 1 dpecka users  138M Mar 26 00:02 DSC_0569-1.tif
-rw-r--r-- 1 dpecka users  138M Mar 26 13:48 DSC_8026.tif
-rw-r--r-- 1 dpecka users  138M Mar 28 23:29 DSC_3793.tif
-rw-r--r-- 1 dpecka users  138M Mar 28 22:22 DSC_3727.tif
-rw-r--r-- 1 dpecka users  138M Mar 28 22:31 DSC_3726.tif
-rw-r--r-- 1 dpecka users  139M Mar 10 23:10 DSCF3098.tif
-rw-r--r-- 1 dpecka users  139M Mar 15 20:22 DSCF3438.tif
-rw-r--r-- 1 dpecka users  139M Mar 15 20:23 DSCF3437.tif
-rw-r--r-- 1 dpecka users  139M Mar 15 21:50 DSCF3431.tif
-rw-r--r-- 1 dpecka users  139M Mar 15 21:50 DSCF3429.tif
-rw-r--r-- 1 dpecka users  139M Mar 15 19:57 DSCF3424.tif
-rw-r--r-- 1 dpecka users  139M Mar 15 19:45 DSCF3423.tif

I believe that the issue is caused when sending the image to external editor (default gimp) .. reproduced with stable 5.8 and also with RawTherapee_dev_5.8-2735-g8bb06c2_20210121.AppImage

suggested resolution: change RT's /tmp directory somewhere to user's home directory ...

thank you for your support and thank you for RT, it's awesome software !!

regards, dan

**Comment by ghost** [comment]
I don't think this is a RawTherapee issue but a Linux (distro specific) issue.

RawTherapee dumps its temporary files in the only sane place it can. 

/tmp should be cleaned automatically by the system when (re)booting and most distro will do this either via rcS when using init or  tmp.conf.d when using systemd or some use plain old cron for this.

Then again, if one doesn't reboot every so often (for whatever reason, although it isn't smart) no files/dirs are removed....

**Comment by dpecka** [comment]
> I don't think this is a RawTherapee issue but a Linux (distro specific) issue.
> 
> RawTherapee dumps its temporary files in the only sane place it can.
> 
> /tmp should be cleaned automatically by the system when (re)booting and most distro will do this either via rcS when using init or tmp.conf.d when using systemd or some use plain old cron for this.
> 
> Then again, if one doesn't reboot every so often (for whatever reason, although it isn't smart) no files/dirs are removed....

Hello Jade ..

let me kindly disagree .. it's common issue and usually the various software smartly use 0077 umask for those files (like everything else aka firefox, thunderbird, ssh, etc ...) or it stores the files in user's home folder ... this is not the decision of operating system that RT saves temporarily files in /tmp with default user's umaks 0022 :D

regards, dan

ps. look on this, just last set of random files from my /tmp ... that DSC*** is RT created

```
> ls -lrt /tmp/ | tail -22
srwxr-xr-x 1 root   root          0 Feb 22 09:43 sddm-authe2cd0213-bb86-4d2c-837d-4e1053dd3da1
srwx------ 1 sddm   sddm          0 Feb 22 09:43 sddm-:0-ZulOzg
drwx------ 1 dpecka users        20 Feb 22 09:44 ssh-ZfUlMUyijxKQ
-rw------- 1 dpecka users       499 Mar  1 09:58 nsmail.sig
-rw------- 1 dpecka users         0 Mar  1 17:59 dolphin.V21769
-rw------- 1 dpecka users         0 Mar  1 18:00 dolphin.y21769
drwx------ 1 dpecka users        96 Mar 18 01:40 MozillaMailnews
drwx------ 1 dpecka users       272 Mar 18 19:43 lu15509f7wrtq.tmp
srwx------ 1 sddm   sddm          0 Mar 20 14:15 sddm--vFXTnQ
drwx------ 1 root   root          6 Mar 20 18:19 systemd-private-4f96d49a4d194bac8252434541e5c572-chronyd.service-yxEees
srwxr-xr-x 1 root   root          0 Mar 20 18:19 sddm-auth1f919f52-5b1b-4c62-bc36-810142b91980
srwx------ 1 sddm   sddm          0 Mar 20 18:19 sddm-:0-dwpnQK
drwx------ 1 dpecka users        20 Mar 20 18:20 ssh-SRRm8OteeYGa
drwx------ 1 root   root          6 Mar 20 18:20 systemd-private-4f96d49a4d194bac8252434541e5c572-rtkit-daemon.service-Lsbu11
drwx------ 1 dpecka users       138 Mar 20 18:20 akonadi-dpecka.jpkitz
drwx------ 1 dpecka users         0 Mar 25 15:33 ark-DtL1wu
drwx------ 1 dpecka users     15072 Mar 26 21:54 mozilla_dpecka0
drwx------ 1 dpecka users         0 Mar 29 10:37 Temp-e6ba2a2d-3ba4-45c5-be01-c31dce365148
drwx------ 1 dpecka users        34 Mar 29 10:37 Temp-7f7d95dd-da86-469e-8875-b6feb04d073e
-rw-r--r-- 1 dpecka users 144868084 Mar 29 10:52 DSCF3074.tif
drwx------ 1 dpecka users         0 Mar 29 11:05 Temp-1e149f99-f4e9-489f-ba2c-5ccc6a92becf
drwx------ 1 dpecka users        34 Mar 29 11:36 Temp-42ac3a4a-4a0f-4a5e-abcd-693ae085d836
```

**Comment by heckflosse** [comment]
> ps. look on this, just last set of random files from my /tmp ... that DSC*** is RT created

the only file created by RT should be `DSCF3074.tif` Though perhaps we could set less permissions...

**Comment by ghost** [comment]
> RT leaves behind in /tmp directory files with 755 permissions, **they could eventually fill the drive**

That is what you wrote/complained about and what I addressed.....

About the umaks: It all depends on the program which mask is used. Krita also creates files with a 022 mask. I would be worried if ssh(d), mail or a system related file would do it.

**Comment by Lawrence37** [comment]
I agree the temporary files should have permissions set to 600. Even better to also place them in a sub-directory with 700 permissions.

**Comment by dpecka** [comment]
> I agree the temporary files should have permissions set to 600. Even better to also place them in a sub-directory with 700 permissions.

the ultimate resolution would be if RT uses some different directory like for exampe ~/.cache .. another problem is, that RT is leaving those files behind ... simply, if you send a file to external editor, RT creates that and doesn't care anymore .. I think that for this kind of transactions between the RT and external editor is the /tmp filesystem a bad choice ... let's make it somewhere in the user's home ... then no need to play with umask and default would be fine ...

**Comment by Thanatomanic** [comment]
Does anybody maybe have an example of how other software handles these kind of temporary files? I think it should be the job of the software to make sure it cleans up after itself tbh.

**Comment by ghost** [comment]
> @dpecka: the ultimate resolution would be if RT uses some different directory like for exampe ~/.cache

I still believe that /tmp is the correct place to put these files. That is assuming that those files are placed in their own (sub)directory and have limited permissions. This isn't the case at the moment as pointed out above.

Using /tmp will take care of the cleaning automatically and thus it is not to the user to have to search their own home (sub)directory structures for possible leftover files. I'm thinking wider than just RT. I don't want to go leftover file/directory hunting every so often 'cause (many) programs have their own tmp directory somewhere in the home directories.

> let's make it somewhere in the user's home ... **then no need to play with umask and default would be fine**

I don't think so. The default umask is 022, which will create files with the readable bit set for the world (and directories as well).

From a privacy/security point of view this _wouldn't_ solve anything. This is the reason why I've set a 0027 umask for my normal users (which comes with its own caveats....).

> @Thanatomanic: think it should be the job of the software to make sure it cleans up after itself tbh.

I agree, but that is not always possible. I'm fighting with the wavelets tab at the moment; Every so often RT (dev) segfaults when using the denoise section. This means that nothing will be cleaned by RT.

This is another reason why /tmp is preferable to a local RT specific location. Even if RT segfaults/crashes it will be cleaned by the system during the next (re)boot.

Just my 2c, as always :smile: 

EDIT: would -> wouldn't

**Comment by dpecka** [comment]
> Does anybody maybe have an example of how other software handles these kind of temporary files? I think it should be the job of the software to make sure it cleans up after itself tbh.

See please my copy-paste from terminal above .. they are often using /tmp with secure permissions (enforced umask 0077) because they expect (as jade-nl noted), that /tmp filesystem is periodically cleaned ...

issue with files left behind is only while opening files in external editor .. it is a good question if RT should forcefully delete a files that it sent to external editor no matter if they are already opened in running instance of external editor ... still there are two levels of issue:

1) insecure permissions of those files (default umask right now, 0077 should be enforced since the /tmp dir is system wide /tmp)
2) files are left behind after closing RT session - this doesn't necessarily mean that they should be deleted .. worth of consideration ...

regards, dan

**Comment by dpecka** [comment]
> I still believe that /tmp is the correct place to put these files. That is assuming that those files are placed in their own (sub)directory and have limited permissions. This isn't the case at the moment as pointed out above.
>  ..
> Using /tmp will take care of the cleaning automatically and thus it is not to the user to have to search their own home (sub)directory structures for possible leftover files. I'm thinking wider than just RT. I don't want to go leftover file/directory hunting every so often 'cause (many) programs have their own tmp directory somewhere in the home directories.

I agree ..


> From a privacy/security point of view this would solve anything. This is the reason why I've set a 0027 umask for my normal users (which comes with its own caveats....).

sounds good ... other programs who use a /tmp enforce 0077, let's do it same

**Comment by da2x** [comment]
Switching to `/tmp/com.rawtherapee/` would enable stricter Flatpak permissions. It could switch to `--filesystem=/tmp/com.rawtherapee:create` instead of full blown `--filesystem=/tmp`. Flatpak’ed GIMP could be given the same permission by the user.

I seems like swapping a couple of `GLib.get_tmp_dir` for [`GLib.Dir.make_tmp("com.rawtherapee")`](https://docs.gtk.org/glib/type_func.Dir.make_tmp.html) should be all that’s needed. The directory is assigned the current user’s permissions, so it solves part of #6358 too.

**Comment by Lawrence37** [comment]
What is the problem to be solved in #6358? The pull request is ready. It's just waiting on the 5.9 release.
